File: pkg/testhelpers/business_case.go
package testhelpers

import (
	"time"

	"github.com/google/uuid"
	"github.com/guregu/null"

	"github.com/cms-enterprise/easi-app/pkg/models"
)

// EstimatedLifecycleCostOptions allows cost options to be customized
type EstimatedLifecycleCostOptions struct {
	BusinessCaseID *uuid.UUID
	Solution       *models.LifecycleCostSolution
	Phase          *models.LifecycleCostPhase
	Year           *models.LifecycleCostYear
	Cost           *int64
}

// NewEstimatedLifecycleCost helps generate a new lifecycle cost for a given Business Case
func NewEstimatedLifecycleCost(opts EstimatedLifecycleCostOptions) models.EstimatedLifecycleCost {
	development := models.LifecycleCostPhaseDEVELOPMENT
	cost := int64(100)
	elc := models.EstimatedLifecycleCost{
		ID:             uuid.New(),
		BusinessCaseID: uuid.New(),
		Solution:       models.LifecycleCostSolutionPREFERRED,
		Phase:          &development,
		Year:           models.LifecycleCostYear1,
		Cost:           &cost,
	}
	if opts.Solution != nil {
		elc.Solution = *opts.Solution
	}
	if opts.Phase != nil {
		elc.Phase = opts.Phase
	}
	if opts.Year != nil {
		elc.Year = *opts.Year
	}
	if opts.Cost != nil {
		elc.Cost = opts.Cost
	}
	if opts.BusinessCaseID != nil {
		elc.BusinessCaseID = *opts.BusinessCaseID
	}
	return elc
}

// NewValidLifecycleCosts helps generate valid lifecycle costs for a given Business Case ready for submission
func NewValidLifecycleCosts(id *uuid.UUID) models.EstimatedLifecycleCosts {
	dev := models.LifecycleCostPhaseDEVELOPMENT
	om := models.LifecycleCostPhaseOPERATIONMAINTENANCE
	other := models.LifecycleCostPhaseOTHER
	cost := int64(100)
	return models.EstimatedLifecycleCosts{
		models.EstimatedLifecycleCost{
			BusinessCaseID: *id,
			Solution:       models.LifecycleCostSolutionPREFERRED,
			Year:           models.LifecycleCostYear1,
			Phase:          &dev,
			Cost:           &cost,
		},
		models.EstimatedLifecycleCost{
			BusinessCaseID: *id,
			Solution:       models.LifecycleCostSolutionPREFERRED,
			Year:           models.LifecycleCostYear1,
			Phase:          &other,
			Cost:           &cost,
		},
		models.EstimatedLifecycleCost{
			BusinessCaseID: *id,
			Solution:       models.LifecycleCostSolutionPREFERRED,
			Year:           models.LifecycleCostYear2,
			Phase:          &om,
			Cost:           &cost,
		},
		models.EstimatedLifecycleCost{
			BusinessCaseID: *id,
			Solution:       models.LifecycleCostSolutionPREFERRED,
			Year:           models.LifecycleCostYear3,
			Phase:          &om,
			Cost:           &cost,
		},
		models.EstimatedLifecycleCost{
			BusinessCaseID: *id,
			Solution:       models.LifecycleCostSolutionPREFERRED,
			Year:           models.LifecycleCostYear4,
			Phase:          &om,
			Cost:           &cost,
		},
		models.EstimatedLifecycleCost{
			BusinessCaseID: *id,
			Solution:       models.LifecycleCostSolutionPREFERRED,
			Year:           models.LifecycleCostYear5,
			Phase:          &om,
			Cost:           &cost,
		},
		models.EstimatedLifecycleCost{
			BusinessCaseID: *id,
			Solution:       models.LifecycleCostSolutionA,
			Year:           models.LifecycleCostYear1,
			Phase:          &dev,
			Cost:           &cost,
		},
		models.EstimatedLifecycleCost{
			BusinessCaseID: *id,
			Solution:       models.LifecycleCostSolutionA,
			Year:           models.LifecycleCostYear1,
			Phase:          &om,
			Cost:           &cost,
		},
		models.EstimatedLifecycleCost{
			BusinessCaseID: *id,
			Solution:       models.LifecycleCostSolutionA,
			Year:           models.LifecycleCostYear2,
			Phase:          &om,
			Cost:           &cost,
		},
		models.EstimatedLifecycleCost{
			BusinessCaseID: *id,
			Solution:       models.LifecycleCostSolutionA,
			Year:           models.LifecycleCostYear3,
			Phase:          &om,
			Cost:           &cost,
		},
		models.EstimatedLifecycleCost{
			BusinessCaseID: *id,
			Solution:       models.LifecycleCostSolutionA,
			Year:           models.LifecycleCostYear4,
			Phase:          &om,
			Cost:           &cost,
		},
		models.EstimatedLifecycleCost{
			BusinessCaseID: *id,
			Solution:       models.LifecycleCostSolutionA,
			Year:           models.LifecycleCostYear5,
			Phase:          &om,
			Cost:           &cost,
		},
		models.EstimatedLifecycleCost{
			BusinessCaseID: *id,
			Solution:       models.LifecycleCostSolutionB,
			Year:           models.LifecycleCostYear1,
			Phase:          &dev,
			Cost:           &cost,
		},
		models.EstimatedLifecycleCost{
			BusinessCaseID: *id,
			Solution:       models.LifecycleCostSolutionB,
			Year:           models.LifecycleCostYear1,
			Phase:          &om,
			Cost:           &cost,
		},
		models.EstimatedLifecycleCost{
			BusinessCaseID: *id,
			Solution:       models.LifecycleCostSolutionB,
			Year:           models.LifecycleCostYear2,
			Phase:          &om,
			Cost:           &cost,
		},
		models.EstimatedLifecycleCost{
			BusinessCaseID: *id,
			Solution:       models.LifecycleCostSolutionB,
			Year:           models.LifecycleCostYear3,
			Phase:          &om,
			Cost:           &cost,
		},
		models.EstimatedLifecycleCost{
			BusinessCaseID: *id,
			Solution:       models.LifecycleCostSolutionB,
			Year:           models.LifecycleCostYear4,
			Phase:          &om,
			Cost:           &cost,
		},
		models.EstimatedLifecycleCost{
			BusinessCaseID: *id,
			Solution:       models.LifecycleCostSolutionB,
			Year:           models.LifecycleCostYear5,
			Phase:          &om,
			Cost:           &cost,
		},
	}
}

// NewBusinessCase allows us to generate a Business Case for tests
func NewBusinessCase(systemIntakeID uuid.UUID) models.BusinessCaseWithCosts {
	now := time.Now().UTC()
	year2 := models.LifecycleCostYear2
	bc := models.BusinessCase{
		ID:                                  uuid.New(),
		EUAUserID:                           RandomEUAID(),
		SystemIntakeID:                      systemIntakeID,
		ProjectName:                         null.StringFrom("Test Project name"),
		ProjectAcronym:                      null.StringFrom("Test Project Acronym"),
		Status:                              models.BusinessCaseStatusOPEN,
		Requester:                           null.StringFrom("Test Requester"),
		RequesterPhoneNumber:                null.StringFrom("6666666666"),
		BusinessOwner:                       null.StringFrom("Test Business Owner"),
		BusinessNeed:                        null.StringFrom("Test Business Need"),
		CollaborationNeeded:                 null.StringFrom("Test Collaboration Needed"),
		CurrentSolutionSummary:              null.StringFrom("Test Current Solution Summary"),
		CMSBenefit:                          null.StringFrom("Test CMS Benefit"),
		PriorityAlignment:                   null.StringFrom("Test Priority Alignment"),
		SuccessIndicators:                   null.StringFrom("Test Success Indicators"),
		ResponseToGRTFeedback:               null.StringFrom("Test Response to GRT Feedback"),
		PreferredTitle:                      null.StringFrom("Test Preferred Title"),
		PreferredSummary:                    null.StringFrom("Test Preferred Summary"),
		PreferredAcquisitionApproach:        null.StringFrom("Test Preferred Acquisition Approach"),
		PreferredTargetContractAwardDate:    &now,
		PreferredTargetCompletionDate:       &now,
		PreferredZeroTrustAlignment:         null.StringFrom("Test Preferred Zero Trust Alignment"),
		PreferredHostingType:                null.StringFrom("none"),
		PreferredHasUI:                      null.StringFrom("YES"),
		PreferredPros:                       null.StringFrom("Test Preferred Pros"),
		PreferredCons:                       null.StringFrom("Test Preferred Cons"),
		PreferredCostSavings:                null.StringFrom("Test Preferred Cost Savings"),
		PreferredWorkforceTrainingReqs:      null.StringFrom("Test Preferred Workforce Training Reqs"),
		AlternativeATitle:                   null.StringFrom("Test Alternative A Title"),
		AlternativeASummary:                 null.StringFrom("Test Alternative A Summary"),
		AlternativeAAcquisitionApproach:     null.StringFrom("Test Alternative A Acquisition Approach"),
		AlternativeATargetContractAwardDate: &now,
		AlternativeATargetCompletionDate:    &now,
		AlternativeAZeroTrustAlignment:      null.StringFrom("Test Alternative A Zero Trust Alignment"),
		AlternativeAHostingType:             null.StringFrom("none"),
		AlternativeAHasUI:                   null.StringFrom("YES"),
		AlternativeAPros:                    null.StringFrom("Test Alternative A Pros"),
		AlternativeACons:                    null.StringFrom("Test Alternative A Cons"),
		AlternativeACostSavings:             null.StringFrom("Test Alternative A Cost Savings"),
		AlternativeAWorkforceTrainingReqs:   null.StringFrom("Test Alternative A Workforce Training Reqs"),
		AlternativeBTitle:                   null.StringFrom("Test Alternative B Title"),
		AlternativeBSummary:                 null.StringFrom("Test Alternative B Summary"),
		AlternativeBAcquisitionApproach:     null.StringFrom("Test Alternative B Acquisition Approach"),
		AlternativeBTargetContractAwardDate: &now,
		AlternativeBTargetCompletionDate:    &now,
		AlternativeBZeroTrustAlignment:      null.StringFrom("Test Alternative B Zero Trust Alignment"),
		AlternativeBHostingType:             null.StringFrom("none"),
		AlternativeBHasUI:                   null.StringFrom("YES"),
		AlternativeBPros:                    null.StringFrom("Test Alternative B Pros"),
		AlternativeBCons:                    null.StringFrom("Test Alternative B Cons"),
		AlternativeBCostSavings:             null.StringFrom("Test Alternative B Cost Savings"),
		AlternativeBWorkforceTrainingReqs:   null.StringFrom("Test Alternative B Workforce Training Reqs"),
		CreatedAt:                           &now,
		UpdatedAt:                           &now,
	}
	return models.BusinessCaseWithCosts{
		BusinessCase: bc,
		LifecycleCostLines: models.EstimatedLifecycleCosts{
			NewEstimatedLifecycleCost(
				EstimatedLifecycleCostOptions{Year: &year2},
			),
			NewEstimatedLifecycleCost(EstimatedLifecycleCostOptions{}),
		},
	}
}

File: cypress/support/businessCase.js
cy.businessCase = {
  generalRequestInfo: {
    fillNonBranchingFields: () => {
      cy.get('#BusinessCase-RequestName')
        .type('Easy Access to System Information')
        .should('have.value', 'Easy Access to System Information');

      cy.get('#BusinessCase-ProjectAcronym')
        .type('EASi')
        .should('have.value', 'EASi');

      cy.get('#BusinessCase-RequesterName')
        .type('Casey Doe')
        .should('have.value', 'Casey Doe');

      cy.get('#BusinessCase-BusinessOwnerName')
        .type('Casey Doe')
        .should('have.value', 'Casey Doe');

      cy.get('#BusinessCase-RequesterPhoneNumber')
        .type('1234567890')
        .should('have.value', '1234567890');
    }
  },
  requestDescription: {
    fillNonBranchingFields: () => {
      cy.get('#BusinessCase-BusinessNeed')
        .type('The quick brown fox jumps over the lazy dog.')
        .should('have.value', 'The quick brown fox jumps over the lazy dog.');

      cy.get('#BusinessCase-CollaborationNeeded')
        .type('The quick brown fox jumps over the lazy dog.')
        .should('have.value', 'The quick brown fox jumps over the lazy dog.');

      cy.get('#BusinessCase-CurrentSolutionSummary')
        .type('The quick brown fox jumps over the lazy dog.')
        .should('have.value', 'The quick brown fox jumps over the lazy dog.');

      cy.get('#BusinessCase-CmsBenefit')
        .type('The quick brown fox jumps over the lazy dog.')
        .should('have.value', 'The quick brown fox jumps over the lazy dog.');

      cy.get('#BusinessCase-PriorityAlignment')
        .type('The quick brown fox jumps over the lazy dog.')
        .should('have.value', 'The quick brown fox jumps over the lazy dog.');

      cy.get('#BusinessCase-SuccessIndicators')
        .type('The quick brown fox jumps over the lazy dog.')
        .should('have.value', 'The quick brown fox jumps over the lazy dog.');

      cy.get('#BusinessCase-ResponseToGRTFeedback')
        .type('The quick brown fox jumps over the lazy dog.')
        .should('have.value', 'The quick brown fox jumps over the lazy dog.');
    }
  },
  preferredSolution: {
    fillAllFields: () => {
      cy.get('#BusinessCase-PreferredSolutionTitle')
        .type('Preferred Solution Title')
        .should('have.value', 'Preferred Solution Title');

      cy.get('#BusinessCase-PreferredSolutionSummary')
        .type('Preferred Solution Summary')
        .should('have.value', 'Preferred Solution Summary');

      cy.get('#BusinessCase-PreferredSolutionAcquisitionApproach')
        .type('Preferred Solution Acquisition approach')
        .should('have.value', 'Preferred Solution Acquisition approach');

      cy.get('#BusinessCase-PreferredSolutionTargetContractAwardDate').type(
        '03/13/2025'
      );

      cy.get('#BusinessCase-PreferredSolutionTargetContractAwardDate').should(
        'have.value',
        '03/13/2025'
      );

      cy.get('#BusinessCase-PreferredSolutionTargetCompletionDate').type(
        '03/15/2025'
      );

      cy.get('#BusinessCase-PreferredSolutionTargetCompletionDate').should(
        'have.value',
        '03/15/2025'
      );

      cy.get('#BusinessCase-PreferredSolutionSecurityNotApproved')
        .check({ force: true })
        .should('be.checked');

      cy.get('#BusinessCase-PreferredSolutionSecurityIsBeingReviewedYes')
        .check({ force: true })
        .should('be.checked');

      cy.get('#BusinessCase-PreferredSolutionZeroTrustAlignment')
        .type('Preferred Solution Zero Trust Alignment')
        .should('have.value', 'Preferred Solution Zero Trust Alignment');

      cy.get('#BusinessCase-PreferredSolutionHostingCloud')
        .check({ force: true })
        .should('be.checked');

      cy.get('#BusinessCase-PreferredSolutionCloudLocation')
        .type('AWS')
        .should('have.value', 'AWS');

      cy.get('#BusinessCase-PreferredSolutionCloudServiceType')
        .type('Saas')
        .should('have.value', 'Saas');

      cy.get('#BusinessCase-PreferredSolutionCloudStrategy')
        .type('Just do it')
        .should('have.value', 'Just do it');

      cy.get('#BusinessCase-PreferredHasUserInferfaceYes')
        .check({ force: true })
        .should('be.checked');

      cy.get('#BusinessCase-PreferredWorkforceTrainingReqs')
        .type('Preferred Workforce Training')
        .should('have.value', 'Preferred Workforce Training');

      cy.get('#BusinessCase-PreferredSolutionPros')
        .type('Preferred Solution Pros')
        .should('have.value', 'Preferred Solution Pros');

      cy.get('#BusinessCase-PreferredSolutionCons')
        .type('Preferred Solution Cons')
        .should('have.value', 'Preferred Solution Cons');

      // Estimated Lifecycle Costs Year 1
      cy.get(
        `#BusinessCase-preferredSolution\\.estimatedLifecycleCost\\.operationsMaintenance\\.years\\.year1`
      )
        .type('5')
        .should('have.value', '5');

      // Estimated Lifecycle Costs Year 2
      cy.get(
        `#BusinessCase-preferredSolution\\.estimatedLifecycleCost\\.development\\.years\\.year2`
      )
        .type('10')
        .should('have.value', '10');

      // Estimated Lifecycle Costs Year 4
      cy.get(
        `#BusinessCase-preferredSolution\\.estimatedLifecycleCost\\.operationsMaintenance\\.years\\.year4`
      )
        .type('20')
        .should('have.value', '20');

      // Estimated Lifecycle Costs Year 5
      cy.get(
        `#BusinessCase-preferredSolution\\.estimatedLifecycleCost\\.development\\.years\\.year5`
      )
        .type('25')
        .should('have.value', '25');

      cy.get('#BusinessCase-PreferredSolutionCostSavings')
        .type('0')
        .should('have.value', '0');
    }
  },
  alternativeASolution: {
    fillAllFields: () => {
      cy.get('#BusinessCase-alternativeATitle')
        .type('Alternative A Title')
        .should('have.value', 'Alternative A Title');

      cy.get('#BusinessCase-alternativeASummary')
        .type('Alternative A Summary')
        .should('have.value', 'Alternative A Summary');

      cy.get('#BusinessCase-alternativeAAcquisitionApproach')
        .type('Alternative A AcquisitionApproach')
        .should('have.value', 'Alternative A AcquisitionApproach');

      cy.get('#BusinessCase-alternativeATargetContractAwardDate').type(
        '03/13/2025'
      );

      cy.get('#BusinessCase-alternativeATargetContractAwardDate').should(
        'have.value',
        '03/13/2025'
      );

      cy.get('#BusinessCase-alternativeATargetCompletionDate').type(
        '03/15/2025'
      );

      cy.get('#BusinessCase-alternativeATargetCompletionDate').should(
        'have.value',
        '03/15/2025'
      );

      cy.get('#BusinessCase-alternativeASecurityNotApproved')
        .check({ force: true })
        .should('be.checked');

      cy.get('#BusinessCase-alternativeASecurityIsBeingReviewedYed')
        .check({ force: true })
        .should('be.checked');

      cy.get('#BusinessCase-alternativeAZeroTrustAlignment')
        .type('Alternative A Zero Trust Alignment')
        .should('have.value', 'Alternative A Zero Trust Alignment');

      cy.get('#BusinessCase-alternativeAHostingCloud')
        .check({ force: true })
        .should('be.checked');

      cy.get('#BusinessCase-alternativeACloudLocation')
        .type('AWS')
        .should('have.value', 'AWS');

      cy.get('#BusinessCase-alternativeACloudServiceType')
        .type('Saas')
        .should('have.value', 'Saas');

      cy.get('#BusinessCase-alternativeACloudStrategy')
        .type('Just do it')
        .should('have.value', 'Just do it');

      cy.get('#BusinessCase-alternativeAHasUserInferfaceYes')
        .check({ force: true })
        .should('be.checked');

      cy.get('#BusinessCase-alternativeAWorkforceTrainingReqs')
        .type('Alternative A Workforce Training')
        .should('have.value', 'Alternative A Workforce Training');

      cy.get('#BusinessCase-alternativeAPros')
        .type('Alternative A Pros')
        .should('have.value', 'Alternative A Pros');

      cy.get('#BusinessCase-alternativeACons')
        .type('Alternative A Cons')
        .should('have.value', 'Alternative A Cons');

      // Estimated Lifecycle Costs Years 1
      cy.get(
        `#BusinessCase-alternativeA\\.estimatedLifecycleCost\\.development\\.years\\.year1`
      )
        .type('2')
        .should('have.value', '2');

      // Estimated Lifecycle Costs Years 2
      cy.get(
        `#BusinessCase-alternativeA\\.estimatedLifecycleCost\\.operationsMaintenance\\.years\\.year2`
      )
        .type('6')
        .should('have.value', '6');

      // Estimated Lifecycle Costs Years 4
      cy.get(
        `#BusinessCase-alternativeA\\.estimatedLifecycleCost\\.operationsMaintenance\\.years\\.year4`
      )
        .type('10')
        .should('have.value', '10');

      // Estimated Lifecycle Costs Years 5
      cy.get(
        `#BusinessCase-alternativeA\\.estimatedLifecycleCost\\.development\\.years\\.year5`
      )
        .type('12')
        .should('have.value', '12');

      cy.get('#BusinessCase-alternativeACostSavings')
        .type('Alternative A Cost Savings')
        .should('have.value', 'Alternative A Cost Savings');
    }
  },
  alternativeBSolution: {}
};
